File: OCastDemoObjC/DetailViewController.m
//
//  DetailViewController.m
//  OCastDemoObjC
//
//  Created by François Suc on 26/06/2019.
//  Copyright © 2019 Orange. All rights reserved.
//

#import "DetailViewController.h"
#import "Constants.h"

@interface DetailViewController ()

@property (weak, nonatomic) IBOutlet UIButton *castButton;
@property (weak, nonatomic) IBOutlet UIButton *stopButton;
@property (weak, nonatomic) IBOutlet UIButton *pauseResumeButton;
@property (weak, nonatomic) IBOutlet UISlider *progressionSlider;
@property (weak, nonatomic) IBOutlet UILabel *startLabel;
@property (weak, nonatomic) IBOutlet UILabel *endLabel;
@property (weak, nonatomic) IBOutlet UISlider *volumeSlider;
@property (weak, nonatomic) IBOutlet UIButton *metadataButton;
@property (weak, nonatomic) IBOutlet UILabel *titleLabel;
@property (weak, nonatomic) IBOutlet UILabel *subtitleLabel;

/// The current playback status.
@property(nonatomic, strong, nullable) MediaPlaybackStatus * currentPlaybackStatus;

/// The time formatter.
@property(nonatomic, strong, nonnull) NSDateComponentsFormatter * timeFormatter;

@end

@implementation DetailViewController

- (void)setCurrentPlaybackStatus:(MediaPlaybackStatus *)currentPlaybackStatus {
    _currentPlaybackStatus = currentPlaybackStatus;
    [self updateUI];
}

- (id)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        self.timeFormatter = [[NSDateComponentsFormatter alloc] init];
        self.timeFormatter.allowedUnits = NSCalendarUnitHour | NSCalendarUnitMinute | NSCalendarUnitSecond;
        self.timeFormatter.zeroFormattingBehavior = NSDateComponentsFormatterZeroFormattingBehaviorPad;
    }
    
    return self;
}

- (void)dealloc {
    [self.device disconnect:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = self.device.friendlyName;
    self.device.applicationName = OCastDemoApplicationName;
    
    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(playbackStatusNotification:)
                                               name:NSNotification.playbackStatusEventNotification
                                             object:self.device];
    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(applicationDidEnterBackground)
                                               name:UIApplicationDidEnterBackgroundNotification
                                             object:nil];
    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(applicationWillEnterForeground)
                                               name:UIApplicationWillEnterForegroundNotification
                                             object:nil];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self updatePlaybackStatus];
}

// MARK: - Private methods

- (void)connect:(void (^)(BOOL))completion {
    SSLConfiguration * sslConfiguration = [[SSLConfiguration alloc] init];
    sslConfiguration.disablesSSLCertificateValidation = YES;
    __weak DetailViewController * weakSelf = self;
    [self.device connect:sslConfiguration completion:^(NSError * _Nullable error) {
        if (error != nil) {
            [weakSelf show:error beforeControllerDismissed:YES];
            completion(NO);
        } else {
            completion(YES);
        }
    }];
}

- (void)show:(NSError *)error beforeControllerDismissed:(BOOL)dismissesController {
    UIAlertController * alertController = [UIAlertController alertControllerWithTitle:@"OCastDemo"
                                                                              message:error.localizedDescription
                                                                       preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction * alertAction = [UIAlertAction actionWithTitle:@"OK"
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction * _Nonnull action) {
                                                             if (dismissesController) {
                                                                 [self.navigationController popToRootViewControllerAnimated:YES];
                                                             }
                                                         }];
    [alertController addAction:alertAction];
    [self presentViewController:alertController animated:YES completion:nil];
}

- (void)ensureConnected:(void (^)(BOOL))completion {
    if (self.device.state != DeviceStateConnected) {
        [self connect:completion];
    } else {
        completion(YES);
    }
}

- (void)updatePlaybackStatus {
    __weak DetailViewController * weakSelf = self;
    [self ensureConnected:^(BOOL connected) {
        if (connected) {
            [weakSelf.device mediaPlaybackStatusWithCompletion:^(MediaPlaybackStatus * _Nullable playbackStatus, NSError * _Nullable error) {
                if (error != nil) {
                    [weakSelf show:error beforeControllerDismissed:NO];
                } else {
                    weakSelf.currentPlaybackStatus = playbackStatus;
                }
            }];
        }
    }];
}

- (void)updateUI {
    if (self.currentPlaybackStatus != nil) {
        switch (self.currentPlaybackStatus.state) {
            case MediaPlaybackStateIdle:
            case MediaPlaybackStateUnknown:
                self.castButton.enabled = YES;
                self.stopButton.enabled = NO;
                self.pauseResumeButton.enabled = NO;
                self.progressionSlider.enabled = NO;
                [self.progressionSlider setValue:0.0 animated:YES];
                self.startLabel.text = @"-";
                self.endLabel.text = @"-";
                self.volumeSlider.enabled = NO;
                [self.volumeSlider setValue:0.0 animated:YES];
                self.metadataButton.enabled = NO;
                self.titleLabel.text = @"-";
                self.subtitleLabel.text = @"-";
                break;
            case MediaPlaybackStateBuffering:
            case MediaPlaybackStatePlaying:
            case MediaPlaybackStatePaused:
                self.castButton.enabled = NO;
                self.stopButton.enabled = YES;
                self.pauseResumeButton.enabled = YES;
                NSString * buttonTitle = self.currentPlaybackStatus.state == MediaPlaybackStatePaused ? @"Resume" : @"Pause";
                [self.pauseResumeButton setTitle:buttonTitle forState:UIControlStateNormal];
                self.progressionSlider.enabled = YES;
                [self.progressionSlider setValue:self.currentPlaybackStatus.position / self.currentPlaybackStatus.duration animated:YES];
                self.startLabel.text = [self.timeFormatter stringFromTimeInterval:self.currentPlaybackStatus.position];
                self.endLabel.text = [self.timeFormatter stringFromTimeInterval:self.currentPlaybackStatus.duration];
                self.volumeSlider.enabled = YES;
                [self.volumeSlider setValue:self.currentPlaybackStatus.volume animated:YES];
                self.metadataButton.enabled = YES;
                break;
        }
    }
}

// MARK: - UI events methods

- (IBAction)castButtonClicked:(id)sender {
    MediaPrepareCommandParams * params = [[MediaPrepareCommandParams alloc] initWithUrl:OCastDemoMovieURLString
                                                                              frequency:1
                                                                                  title:@"Movie Sample"
                                                                               subtitle:@"OCast"
                                                                                   logo:@""
                                                                              mediaType:MediaTypeVideo
                                                                           transferMode:MediaTransferModeBuffered
                                                                               autoPlay:YES];
    __weak DetailViewController * weakSelf = self;
    [self ensureConnected:^(BOOL connected) {
        if (connected) {
            [weakSelf.device prepareMedia:params withOptions:nil completion:^(NSError * _Nullable error) {
                if (error != nil) {
                    [weakSelf show:error beforeControllerDismissed:NO];
                }
            }];
        }
    }];
}

- (IBAction)stopButtonClicked:(id)sender {
    __weak DetailViewController * weakSelf = self;
    [self ensureConnected:^(BOOL connected) {
        if (connected) {
            [weakSelf.device stopMediaWithCompletion:^(NSError * _Nullable error) {
                if (error != nil) {
                    [weakSelf show:error beforeControllerDismissed:NO];
                }
            }];
        }
    }];
}

- (IBAction)pauseResumeButtonClicked:(id)sender {
    __weak DetailViewController * weakSelf = self;
    [self ensureConnected:^(BOOL connected) {
        if (connected) {
            if (weakSelf.currentPlaybackStatus.state == MediaPlaybackStatePaused) {
                [weakSelf.device resumeMediaWithCompletion:^(NSError * _Nullable error) {
                    if (error != nil) {
                        [weakSelf show:error beforeControllerDismissed:NO];
                    }
                }];
            } else {
                [weakSelf.device pauseMediaWithCompletion:^(NSError * _Nullable error) {
                    if (error != nil) {
                        [weakSelf show:error beforeControllerDismissed:NO];
                    }
                }];
            }
        }
    }];
}

- (IBAction)progressionSliderChanged:(id)sender {
    if (self.currentPlaybackStatus != nil) {
        __weak DetailViewController * weakSelf = self;
        [self ensureConnected:^(BOOL connected) {
            if (connected) {
                double position = weakSelf.progressionSlider.value * weakSelf.currentPlaybackStatus.duration;
                [weakSelf.device seekMediaTo:position completion:^(NSError * _Nullable error) {
                    if (error != nil) {
                        [weakSelf show:error beforeControllerDismissed:NO];
                    }
                }];
            }
        }];
    }
}
- (IBAction)volumeSliderChanged:(id)sender {
    __weak DetailViewController * weakSelf = self;
    [self ensureConnected:^(BOOL connected) {
        if (connected) {
            [weakSelf.device setMediaVolume:weakSelf.volumeSlider.value completion:^(NSError * _Nullable error) {
                if (error != nil) {
                    [weakSelf show:error beforeControllerDismissed:NO];
                }
            }];
        }
    }];
}

- (IBAction)metadataButtonClicked:(id)sender {
    __weak DetailViewController * weakSelf = self;
    [self ensureConnected:^(BOOL connected) {
        if (connected) {
            [self.device mediaMetadataWithCompletion:^(MediaMetadata * _Nullable metadata, NSError * _Nullable error) {
                if (error != nil) {
                    [weakSelf show:error beforeControllerDismissed:NO];
                } else {
                    weakSelf.titleLabel.text = [NSString stringWithFormat:@"Titre: %@", metadata.title];
                    weakSelf.subtitleLabel.text = [NSString stringWithFormat:@"Sous-titre: %@", metadata.subtitle];
                }
            }];
        }
    }];
}

// MARK: - Notifications

- (void)playbackStatusNotification:(NSNotification *)notification {
    self.currentPlaybackStatus = notification.userInfo[DeviceUserInfoKey.playbackStatusUserInfoKey];
}

- (void)applicationDidEnterBackground {
    [self.device disconnect:nil];
}

- (void)applicationWillEnterForeground {
    [self updatePlaybackStatus];
}

@end

File: OCastDemoSwift/DetailViewController.swift
//
// ViewController.swift
//
// Copyright 2018 Orange
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
//

import OCast
import UIKit

class DetailViewController: UIViewController {

    @IBOutlet weak var castButton: UIButton!
    @IBOutlet weak var stopButton: UIButton!
    @IBOutlet weak var pauseResumeButton: UIButton!
    @IBOutlet weak var progressionSlider: UISlider!
    @IBOutlet weak var startLabel: UILabel!
    @IBOutlet weak var endLabel: UILabel!
    @IBOutlet weak var volumeSlider: UISlider!
    @IBOutlet weak var metadataButton: UIButton!
    @IBOutlet weak var titleLabel: UILabel!
    @IBOutlet weak var subtitleLabel: UILabel!
    
    /// The devivce.
    var device: Device!
    
    /// The current playback status.
    private var currentPlaybackStatus: MediaPlaybackStatus? {
        didSet {
            updateUI()
        }
    }
    
    /// The time formatter.
    private let timeFormatter = DateComponentsFormatter()
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        timeFormatter.allowedUnits = [.hour, .minute, .second]
        timeFormatter.zeroFormattingBehavior = .pad
    }
    
    deinit {
        device.disconnect()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = device.friendlyName
        device.applicationName = OCastDemoApplicationName
        
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(playbackStatusNotification),
                                               name: .playbackStatusEventNotification,
                                               object: device)
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(applicationDidEnterBackground),
                                               name: UIApplication.didEnterBackgroundNotification,
                                               object: nil)
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(applicationWillEnterForeground),
                                               name: UIApplication.willEnterForegroundNotification,
                                               object: nil)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        updatePlaybackStatus()
    }
    
    // MARK: - Private methods
    
    private func connect(completion: @escaping (Bool) -> Void) {
        let sslConfiguration = SSLConfiguration()
        sslConfiguration.disablesSSLCertificateValidation = true
        device.connect(sslConfiguration) { [weak self] error in
            if let error = error {
                self?.show(error, beforeControllerDismissed: true)
                completion(false)
            } else {
                completion(true)
            }
        }
    }
    
    private func show(_ error: Error, beforeControllerDismissed: Bool = false) {
        let alertController = UIAlertController(title: "OCastDemo", message: error.localizedDescription, preferredStyle: .alert)
        let alertAction = UIAlertAction(title: "OK", style: .default, handler: { _ in
            if beforeControllerDismissed {
                self.navigationController?.popToRootViewController(animated: true)
            }
        })
        alertController.addAction(alertAction)
        present(alertController, animated: true, completion: nil)
    }
    
    private func ensureConnected(completion: @escaping (Bool) -> Void) {
        if device.state != .connected {
            connect(completion: completion)
        } else {
            completion(true)
        }
    }
    
    private func updatePlaybackStatus() {
        ensureConnected { [weak self] connected in
            if connected {
                self?.device.playbackStatus(completion: { playbackStatus, error in
                    if let error = error {
                        self?.show(error)
                    } else {
                        self?.currentPlaybackStatus = playbackStatus
                    }
                })
            }
        }
    }
    
    private func updateUI() {
        guard let currentPlaybackStatus = currentPlaybackStatus else { return }
        
        switch currentPlaybackStatus.state {
        case .idle, .unknown:
            castButton.isEnabled = true
            stopButton.isEnabled = false
            pauseResumeButton.isEnabled = false
            progressionSlider.isEnabled = false
            progressionSlider.setValue(0.0, animated: true)
            startLabel.text = "-"
            endLabel.text = "-"
            volumeSlider.isEnabled = false
            volumeSlider.setValue(0.0, animated: true)
            metadataButton.isEnabled = false
            titleLabel.text = "-"
            subtitleLabel.text = "-"
        case .buffering, .playing, .paused:
            castButton.isEnabled = false
            stopButton.isEnabled = true
            pauseResumeButton.isEnabled = true
            pauseResumeButton.setTitle(currentPlaybackStatus.state == .paused ? "Resume" : "Pause", for: .normal)
            progressionSlider.isEnabled = true
            progressionSlider.setValue(Float(currentPlaybackStatus.position / currentPlaybackStatus.duration), animated: true)
            startLabel.text = timeFormatter.string(from: currentPlaybackStatus.position) ?? "-"
            endLabel.text = timeFormatter.string(from: currentPlaybackStatus.duration) ?? "-"
            volumeSlider.isEnabled = true
            volumeSlider.setValue(currentPlaybackStatus.volume, animated: true)
            metadataButton.isEnabled = true
        }
    }
    
    // MARK: - UI events methods
    
    @IBAction func castButtonClicked(_ sender: Any) {
        let mediaPrepareCommand = MediaPrepareCommand(url: OCastDemoMovieURLString,
                                                      frequency: 1,
                                                      title: "Movie Sample",
                                                      subtitle: "OCast",
                                                      logo: "",
                                                      mediaType: .video,
                                                      transferMode: .buffered,
                                                      autoPlay: true)
        ensureConnected { [weak self] connected in
            if connected {
                self?.device.prepare(mediaPrepareCommand, completion: { error in
                    if let error = error {
                        self?.show(error)
                    }
                })
            }
        }
    }
    
    @IBAction func stopButtonClicked(_ sender: Any) {
        ensureConnected { [weak self] connected in
            if connected {
                self?.device.stop(completion: { error in
                    if let error = error {
                        self?.show(error)
                    }
                })
            }
        }
    }
    
    @IBAction func pauseResumeButtonClicked(_ sender: Any) {
        ensureConnected { [weak self] connected in
            if connected {
                if self?.currentPlaybackStatus?.state == .paused {
                    self?.device.resume(completion: { error in
                        if let error = error {
                            self?.show(error)
                        }
                    })
                } else {
                    self?.device.pause(completion: { error in
                        if let error = error {
                            self?.show(error)
                        }
                    })
                }
            }
        }
    }
    
    @IBAction func progressionSliderChanged(_ sender: Any) {
        guard let currentPlaybackStatus = currentPlaybackStatus else { return }
        
        ensureConnected { [weak self] connected in
            guard let `self` = self else { return }
            if connected {
                let position = Double(self.progressionSlider.value) * currentPlaybackStatus.duration
                self.device.seek(to: position, completion: { error in
                    if let error = error {
                        self.show(error)
                    }
                })
            }
        }
    }
    @IBAction func volumeSliderChanged(_ sender: Any) {
        ensureConnected { [weak self] connected in
            guard let `self` = self else { return }
            if connected {
                self.device.setVolume(self.volumeSlider.value, completion: { error in
                    if let error = error {
                        self.show(error)
                    }
                })
            }
        }
    }
    
    @IBAction func metadataButtonClicked(_ sender: Any) {
        ensureConnected { [weak self] connected in
            if connected {
                self?.device.metadata(completion: { metadata, error in
                    if let error = error {
                        self?.show(error)
                    } else if let metadata = metadata {
                        self?.titleLabel.text = "Titre: \(metadata.title)"
                        self?.subtitleLabel.text = "Sous-titre: \(metadata.subtitle)"
                    }
                })
            }
        }
    }
    
    // MARK: - Notifications
    
    @objc func playbackStatusNotification(_ notification: Notification) {
        currentPlaybackStatus = notification.userInfo?[DeviceUserInfoKey.playbackStatusUserInfoKey] as? MediaPlaybackStatus
    }
    
    @objc func applicationDidEnterBackground() {
        device.disconnect()
    }
    
    @objc func applicationWillEnterForeground() {
        updatePlaybackStatus()
    }
}
